Solutions/Marain.UserNotifications.Abstractions/Marain/UserNotifications/UserNotificationExtensions.cs

```diff
@@ -0,0 +1,146 @@
+﻿// <copyright file="UserNotificationExtensions.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Marain.UserNotifications
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+
+    /// <summary>
+    /// Extension methods for the <see cref="UserNotification"/> class.
+    /// </summary>
+    public static class UserNotificationExtensions
+    {
+        /// <summary>
+        /// Gets the current <see cref="UserNotificationDeliveryStatus"/> for the specified delivery channel.
+        /// </summary>
+        /// <param name="userNotification">The user notification to check.</param>
+        /// <param name="deliveryChannelId">The Id of the delivery channel to retrieve the status for.</param>
+        /// <returns>
+        /// The delivery status for the channel. If not explicitly set for the notification, the value
+        /// <see cref="UserNotificationDeliveryStatus.Unknown"/> is returned.
+        /// </returns>
+        public static UserNotificationDeliveryStatus GetDeliveryStatusForChannel(
+            this UserNotification userNotification,
+            string deliveryChannelId)
+        {
+            if (userNotification is null)
+            {
+                throw new ArgumentNullException(nameof(userNotification));
+            }
+
+            if (string.IsNullOrEmpty(deliveryChannelId))
+            {
+                throw new ArgumentNullException(nameof(deliveryChannelId));
+            }
+
+            return userNotification.ChannelStatuses
+                .SingleOrDefault(x => x.DeliveryChannelId == deliveryChannelId)
+                ?.DeliveryStatus ?? UserNotificationDeliveryStatus.Unknown;
+        }
+
+        /// <summary>
+        /// Gets the current <see cref="UserNotificationReadStatus"/> for the specified delivery channel.
+        /// </summary>
+        /// <param name="userNotification">The user notification to check.</param>
+        /// <param name="deliveryChannelId">The Id of the delivery channel to retrieve the status for.</param>
+        /// <returns>
+        /// The delivery status for the channel. If not explicitly set for the notification, the value
+        /// <see cref="UserNotificationReadStatus.Unknown"/> is returned.
+        /// </returns>
+        public static UserNotificationReadStatus GetReadStatusForChannel(
+            this UserNotification userNotification,
+            string deliveryChannelId)
+        {
```

Comment:
This is so similar to the previous method that I kind of want a thing that gets as far as the `SingleOrDefault`, so that this and the preceding method then essentially reduce down to something more like `=> this.GetStatusForChannel(deliveryChannelId)?.ReadStatus ?? UserNotificationReadStatus.Unknown`

My main reason for that is that when two methods do very similar things I find it easier to read if you put the bits that are the same in a shared method, because it makes the bits that are different easier to spot. (However, reasonable people may disagree on whether this is worth the effort of adding the helper method, so it's not a hill I'd die on.)

Reply:
Agreed